### DISCUSSION TITLE
Strange problem with rest response during login with credentials

### DISCUSSION THREAD

**Discussion by oskar-gmerek** [nextauthjs/next-auth#2210] [Help]
Hello,
I have strange problem and I not even sure if this is next-auth related but I will be very grateful if someone will help me with it or just let me know with what that problem is. I was already made question on the stackoverflow but unfortunately without any answer.

https://stackoverflow.com/questions/68010914/next-auth-when-user-object-have-too-many-items-then-session-request-is-without

In short. Everything is working good unless If rest response is just with a little bit more data. Then the client is getting empty response. I think it maybe problem related to next-auth because when I make the same in postman then everything is fine. Can somebody help me with that?

Regards,
Oskar

**Comment by balazsorban44** [comment]
without your code it's hard to guess, but maybe because of the browser's cookie size limitation, since you are talking about a "bit more data"

https://next-auth.js.org/faq#what-are-the-disadvantages-of-json-web-tokens

you might have to persist data in a database.

https://next-auth.js.org/adapters/overview

**Reply by oskar-gmerek** [reply]
@balazsorban44 Thank you so much!
That's the problem. I just check this and I can see that JWT Token contain all that data what makes it very large (4,5kb what excess that limit). So that's must be the problem.

I'm not a professional developer and I'm not sure if I thinking in a good way.
When I hit the API endpoint for login user then I got whole user object (its to big for cookie), I no need whole object of user to save in that cookie. I want there only most important information like jwt, user id, user name. I can't change that in the backend. 

As I want to use GraphQL in my project (with Apollo Client) I think I have three ways to sort out this.

1. First is to do that what you suggest to me, but is for sure to hard for me as I'm not a proffesional developer.
2. Second way, and probably the best will be to use graphql mutation in [...nextauth].js, but I have problems with to do that as is working in different way on the server side and my attempts to implement it is ending with errors, Can't find good docs for that.
3. Third way is to extract data before is it store to cookie. - That should be the easiest way I think, but I trying to do that from yesterday and always brake something. 

The first way I think is too hard to me so I didn't even try.

The second way I tries like that:
```
...
import {ApolloClient,
  createHttpLink,
  InMemoryCache, gql, mutation} from '@apollo/client'

const ME = gql`
mutation LoginMutation(
  $identifier: String!, $password: String!
) {
  login(
    input: {
      data: {
        identifier: $username, password: $password
      }
    }
  ) {
    jwt 
    user {
        id
      username
      role {
        name
      }
    }
  }
  }
`;
...
```

```
...
      authorize: async (credentials) => {
        const client = new ApolloClient({
          ssrMode: true,
          link: createHttpLink({
            uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
            credentials: 'include',
          }),
          cache: new InMemoryCache(),
        });
        try {
          await client.resetStore();
          const { data: { login: { user, jwt } } } = await client.mutation({
              mutation: ME,
              // variables: { identifier: credentials.username, password: credentials.password }
            });

          if (user) {
              return user;
          }

        } catch (error) {
          const errorMessage = error.response.data.message[0].messages[0].message
          throw new Error(errorMessage)
        }
  }
})
],
...
```
But that is ending with error 'mutation' is not a function.




The third way I was trying like that:
```
...
authorize: async (credentials) => {
        try {
          const full_user = await axios.post(`${process.env.NEXT_PUBLIC_API_URL}/auth/local`, {
            identifier: credentials.username,
            password: credentials.password,
          });
          
          console.log(full_user.data) // <--- that log full user object in my console
          console.log(full_user.data.jwt) // <--- that log jwt in my console
          console.log(full_user.data.user.id) // <--- that log user id in my console
          if (full_user.data) {
            user = {id: full_user.data.user.id, username: full_user.data.user.username, jwt: full_user.data.jwt}
            return user
          } else {
            return null
          } 
        } catch (error) {
          const errorMessage = error.response.data.message[0].messages[0].message
          throw new Error(errorMessage)
        }
      } 
    }),
  ],

...

session: {
    jwt: true,
  },
  callbacks: {
    jwt: async (token, user) => {
      if (user) {
        token.jwt = user.jwt;
        token.user = user
      }
      return Promise.resolve(token);
    },
    session: async (session, token, user) => {
      session.jwt = token.jwt;
      session.user = user;
      console.log(session.jwt)
      return Promise.resolve(session);
      
    },
  },
...

```

Thats ending with error: Cannot read property 'data' of undefined. I don't know how to deal with it.







I will be very grateful for any sugestions. Maybe I tried something close to be good but make anyway simple mistake?

Thank you for your time Balázs

if you will need any other part of code let me know and I will post it. But the rest is simply 'basic' and I think that is the most related part.

**Reply by balazsorban44** [reply]
I suggest you check the docs, and cut down the size, especially if you don't need all the data returned from your provider. (sounds most similar to your 3. alternative) 

https://next-auth.js.org/configuration/callbacks#jwt-callback

https://next-auth.js.org/configuration/callbacks#session-callback

**Reply by oskar-gmerek** [reply]
@balazsorban44 

Thank you. Now I know where I need to work out the solution


I cut it in this way:
```
  callbacks: {
    jwt: async (token, user) => {
      if (user) {
        token.jwt = user.jwt;
        token.user = [
          token.id=user.user.id, 
          token.username=user.user.username
        ]
      }
      return Promise.resolve(token);
    },
    session: async (session, token) => {
      session.jwt = token.jwt;
      session.user = [
        ['id', token.id], 
        ['username', token.username]
      ]
      return Promise.resolve(session);
    },
  },
```


But still can't access to for example an users id by session.user.id. What I'm doing wrong?

**Reply by oskar-gmerek** [reply]
But not in full how I want it. With that solution I don't have access to username by using session.user.username.

I was tried to change 'username' to session.user.username but it also doesn't work.

**Reply by balazsorban44** [reply]
You are mixing arrays and objects.

I suggest you try reading up on JavaScript basics before diving into a bigger project.

**Reply by oskar-gmerek** [reply]
> You are mixing arrays and objects.
> 
> I suggest you try reading up on JavaScript basics before diving into a bigger project.

@balazsorban44 So if I understand good. That should work as expected for me:

```
callbacks: {
    jwt: async (token, user) => {
      if (user) {
        token.jwt = user.jwt;
        token.user = {
          id: user.user.id, 
          username: user.user.username
        }
      }
      return Promise.resolve(token);
    },
    session: async (session, token) => {
      session.jwt = token.jwt;
      session.user = {
        id: token.user.id,
        username: token.user.username
      }
      return Promise.resolve(session);
    },
  },

```

**Reply by oskar-gmerek** [reply]
Great. Work as expected. It was first what I was checking but I mistook ":" with "=" and the errors descriptions in VS take me to that strange place.

It's hard for me to learn from simple examples, when the code is simple everything seems simple. As the code gets more elaborate, these previous simple things become less clear, especially if some code has already been written by a third party.

It seems to me that it is best to learn from real projects, although there are problems like here, basic ones, but with the help of a person like you, I already know that this basic knowledge in this case actually works, and my problems were due to stupid mistake (but doesn't it take a lot of practice not to make mistakes?)

@balazsorban44  Thank you so much for guiding me on the right path to solving this problem.

**Reply by balazsorban44** [reply]
glad that it worked out! I can understand that learning from real projects sounds great, and by all means if you feel that is what works for you, just do it. But if you take an advice, I would find a good javascript intro course just to cover the very basic stuff. 😊
